src/ripple/app/tx/impl/InvariantCheck.cpp

```diff
@@ -0,0 +1,130 @@
+  Copyright (c) 2012-2016 Ripple Labs Inc.
+
+  Permission to use, copy, modify, and/or distribute this software for any
+  purpose  with  or without fee is hereby granted, provided that the above
+  copyright notice and this permission notice appear in all copies.
+
+  THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+  WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+  MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+  ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+  WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+  ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/tx/impl/InvariantCheck.h>
+#include <ripple/basics/Log.h>
+
+namespace ripple {
+
+void
+XRPNotCreated::visitEntry(
+    uint256 const& /*index*/,
+    bool isDelete,
+    std::shared_ptr <SLE const> const& before,
+    std::shared_ptr <SLE const> const& after)
+{
+    if(before)
+    {
+        switch (before->getType())
+        {
+        case ltACCOUNT_ROOT:
+            drops_ -= (*before)[sfBalance].xrp().drops();
+            break;
+        case ltPAYCHAN:
+            drops_ -= ((*before)[sfAmount] - (*before)[sfBalance]).xrp().drops();
+            break;
+        case ltESCROW:
+            drops_ -= (*before)[sfAmount].xrp().drops();
+            break;
+        default:
+            break;
+        }
+    }
+
+    if(after)
+    {
+        switch (after->getType())
+        {
+        case ltACCOUNT_ROOT:
+            drops_ += (*after)[sfBalance].xrp().drops();
+            break;
+        case ltPAYCHAN:
+            if (! isDelete)
+                drops_ += ((*after)[sfAmount] - (*after)[sfBalance]).xrp().drops();
+            break;
+        case ltESCROW:
+            if (! isDelete)
+                drops_ += (*after)[sfAmount].xrp().drops();
+            break;
+        default:
+            break;
+        }
+    }
+}
+
+bool
+XRPNotCreated::finalize(STTx const& tx, TER /*tec*/, beast::Journal const& j)
+{
+    auto fee = tx.getFieldAmount(sfFee).xrp().drops();
+    if(-1*fee <= drops_ && drops_ <= 0)
+        return true;
+
+    JLOG(j.fatal()) << "Invariant failed: XRP net change was " << drops_ <<
+        " on a fee of " << fee;
+    return false;
+}
+
+//------------------------------------------------------------------------------
+
+void
+AccountRootsNotDeleted::visitEntry(
+    uint256 const& /*index*/,
+    bool isDelete,
+    std::shared_ptr <SLE const> const& before,
+    std::shared_ptr <SLE const> const& /*after*/)
+{
+    if (isDelete && before && before->getType() == ltACCOUNT_ROOT)
+        accountDeleted_ = true;
+}
+
+bool
+AccountRootsNotDeleted::finalize(STTx const& /*tx*/, TER /*tec*/, beast::Journal const& j)
```

Comment:
Not to nitpick, but I'd rather not have the commented-out names here (and other places) - they add very little, beyond what the types provide. Having `(STTx const&, TER, beast::Journal const& j)` should be fine. Unless others feel strongly about them staying in, I'd remove them.